File: data.structures/haskell/Demos/Queue/QueuesPerformance.hs
-------------------------------------------------------------------------------
-- Experimental test for evaluating Queues performance
--
-- Data Structures. Grado en Informática. UMA.
-- Pepe Gallardo, 2012
-------------------------------------------------------------------------------

module Demos.Queue.QueuesPerformance where

import DataStructures.Util.Random
import DataStructures.Queue.TwoListsQueue -- LinearQueue
import System.CPUTime

data Operation = Enqueue | Dequeue

-- on average, do 2 enqueues for each dequeue
randomOperations :: Seed -> [Operation]
randomOperations s = randomsIn [Enqueue, Enqueue, Dequeue] s

-- forces queue evaluation by summing its elements
sumQ :: (Num a) => Queue a -> a
sumQ q
 | isEmpty q = 0
 | otherwise = first q + sumQ (dequeue q)

test :: Seed -> Int -> Int
test s n = sumQ (foldr simulate empty (take n (randomOperations s)))

simulate :: Operation -> Queue Int -> Queue Int
simulate Enqueue q = enqueue 0 q
simulate Dequeue q = if isEmpty q then q else dequeue q

main = do
 let tests = 10
 let numOperations = 10000
 t0 <- getCPUTime
 let xs = [ test s numOperations | s <- [0..tests-1]]
 print (sum xs) -- force evaluation
 t1 <- getCPUTime
 let average = toSecs (t1-t0) / fromIntegral tests
 putStrLn ("Tests took "++ show average ++ " secs on average")

toSecs :: Integer -> Double
toSecs x = fromIntegral x / 10^12

File: data.structures/java/demos/queue/QueuesPerformance.java
/**
 * @author Pepe Gallardo, Data Structures, Grado en Informática. UMA.
 *
 * Experimental comparison for queue implementations
 */

package demos.queue;

import dataStructures.queue.Queue;
import dataStructures.queue.ArrayQueue;
import dataStructures.queue.LinkedQueue;

public class QueuesPerformance {
	
	public enum Implementation { WithArray, Linked};
	
	public static double test(Implementation impl, int seed, int length) {
		long t0 = System.currentTimeMillis();
		Queue<Integer> queue = impl==Implementation.WithArray ? new ArrayQueue<>()
				                                              : new LinkedQueue<>(); 
		RandomQueue.fill(queue, seed, length);
		long t1 = System.currentTimeMillis();
		return (t1-t0) / 10e3; //execution time in seconds
	}	
	
	static double avgTime(Implementation impl, int tests) {
		double t = 0.0;
		
		for(int s=0; s<tests; s++) 
			t += test(impl, s, 10000000);

		return t/tests;	
	}
	
	public static void main(String[] args) {
		
		final int tests = 10;
		
		double t1 = avgTime(Implementation.Linked, tests);
		double t2 = avgTime(Implementation.WithArray, tests);
		
		System.out.printf("Average execution time for LinkedQueue is %f seconds\n", t1);
		System.out.printf("Average execution time for ArrayQueue is %f seconds\n", t2);
		System.out.printf("ArrayQueue is %.2f times faster", t1/t2);		
		
	}	
}